avm/src/vm/opcode.rs

```diff
@@ -1522,6 +1522,236 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
+    let out = a + b;
+    let num = i64::from_ne_bytes(out.to_ne_bytes());
+    hand_mem.write_fixed(args[2], num)?;
+    Ok(None)
+  });
+
+  cpu!(ssubi8 => fn(args, hand_mem) {
+    let a = hand_mem.read_fixed(args[0])? as i8;
+    let b = hand_mem.read_fixed(args[1])? as i8;
+    hand_mem.write_fixed(args[2], a.saturating_sub(b) as i64)?;
+    Ok(None)
+  });
+  cpu!(ssubi16 => fn(args, hand_mem) {
+    let a = hand_mem.read_fixed(args[0])? as i16;
+    let b = hand_mem.read_fixed(args[1])? as i16;
+    hand_mem.write_fixed(args[2], a.saturating_sub(b) as i64)?;
+    Ok(None)
+  });
+  cpu!(ssubi32 => fn(args, hand_mem) {
+    let a = hand_mem.read_fixed(args[0])? as i32;
+    let b = hand_mem.read_fixed(args[1])? as i32;
+    hand_mem.write_fixed(args[2], a.saturating_sub(b) as i64)?;
+    Ok(None)
+  });
+  cpu!(ssubi64 => fn(args, hand_mem) {
+    let a = hand_mem.read_fixed(args[0])?;
+    let b = hand_mem.read_fixed(args[1])?;
+    hand_mem.write_fixed(args[2], a.saturating_sub(b))?;
+    Ok(None)
+  });
+  cpu!(ssubf32 => fn(args, hand_mem) {
+    let a = f32::from_ne_bytes((hand_mem.read_fixed(args[0])? as i32).to_ne_bytes());
+    let b = f32::from_ne_bytes((hand_mem.read_fixed(args[1])? as i32).to_ne_bytes());
+    let out = a - b;
+    let num = i32::from_ne_bytes(out.to_ne_bytes()) as i64;
+    hand_mem.write_fixed(args[2], num)?;
+    Ok(None)
+  });
+  cpu!(ssubf64 => fn(args, hand_mem) {
+    let a = f64::from_ne_bytes(hand_mem.read_fixed(args[0])?.to_ne_bytes());
+    let b = f64::from_ne_bytes(hand_mem.read_fixed(args[1])?.to_ne_bytes());
+    let out = a - b;
+    let num = i64::from_ne_bytes(out.to_ne_bytes());
+    hand_mem.write_fixed(args[2], num)?;
+    Ok(None)
+  });
+
+  cpu!(smuli8 => fn(args, hand_mem) {
+    let a = hand_mem.read_fixed(args[0])? as i8;
+    let b = hand_mem.read_fixed(args[1])? as i8;
+    hand_mem.write_fixed(args[2], a.saturating_mul(b) as i64)?;
+    Ok(None)
+  });
+  cpu!(smuli16 => fn(args, hand_mem) {
+    let a = hand_mem.read_fixed(args[0])? as i16;
+    let b = hand_mem.read_fixed(args[1])? as i16;
+    hand_mem.write_fixed(args[2], a.saturating_mul(b) as i64)?;
+    Ok(None)
+  });
+  cpu!(smuli32 => fn(args, hand_mem) {
+    let a = hand_mem.read_fixed(args[0])? as i32;
+    let b = hand_mem.read_fixed(args[1])? as i32;
+    hand_mem.write_fixed(args[2], a.saturating_mul(b) as i64)?;
+    Ok(None)
+  });
+  cpu!(smuli64 => fn(args, hand_mem) {
+    let a = hand_mem.read_fixed(args[0])?;
+    let b = hand_mem.read_fixed(args[1])?;
+    hand_mem.write_fixed(args[2], a.saturating_mul(b))?;
+    Ok(None)
+  });
+  cpu!(smulf32 => fn(args, hand_mem) {
+    let a = f32::from_ne_bytes((hand_mem.read_fixed(args[0])? as i32).to_ne_bytes());
+    let b = f32::from_ne_bytes((hand_mem.read_fixed(args[1])? as i32).to_ne_bytes());
+    let out = a * b;
+    let num = i32::from_ne_bytes(out.to_ne_bytes()) as i64;
+    hand_mem.write_fixed(args[2], num)?;
+    Ok(None)
+  });
```

Comment:
does it make sense to have saturating versions for floats since they're not at all different from the original versions?

Reply:
They are different, though? The "normal" floating point math opcodes actually simulate the integer error failures -- hitting `Infinity` or `Div-By-Zero` causes an Error to be returned instead of a Result-wrapped float.

Reply:
ah, you're right - my bad